### PR TITLE
reject publishing actions with a retriable error code if a earlier task is still publishing

### PR DESCRIPTION
**Pull request by georgew5656**
QOL improvment around segment publishing when multiple tasks in the same task group are pending completion.

### Description
It's possible for the following timeline to occur for a single supervisor.

Task 1 goes into publishing, Task 2 is started
Supervisor is updated or handoff is triggerred manually
Task 2 goes into publishing and immediately publishes (it has not read many events).
Task 2 fails because the offsets in the DB do not match (Task 1 has not completed its SegmentTransactionalInsertAction yet).

Normally task 2 will fail with one of the below errors:
`The new start metadata state[%s] is ahead of the last committed"
          + " end state[%s]. Try resetting the supervisor`

`Inconsistency between stored metadata state[%s] and target state[%s]. Try resetting the supervisor.`

Eventually Task 1 will publish and the supervisor will get itself sorted out but this can lead to several minutes of downtime and lag (and many failed tasks) as the supervisor tries to recover.

This fix attempts to address this issue by having the SegmentTransactionalInsertAction check whether a task can actually publish before trying.

#### Added a forbidden-apis entry ...

The general approach here is to introduce a canPublishSegments method on each supervisor, and then call this method each time a streaming task wants to publish. If this method returns false, then throw a 503 error that the task will then retry.

A alternative approach might be to only try this as a fallback (right now this logic has to be run on every publish). But I thought this might introduce consistency bugs since part of the publish logic would have ran and failed and the logic introduced here is relatively simple (checking stuff in memory). I can update the logic to be exception handling only if others have strong opinions.

The canPublishSegments logic is only implemented on SeekableStreamSupervisor and essentially looks at pendingCompletionTaskGroups to see if there is already a task publishing (if the task is in pendingCompletionTaskGroups, that indicates it is already publishing) with earlier offsets then the task that is trying to publish.

I'm still working on some unit tests for this code.

#### Release note
QOL improvment for resubmitting supervisors or for when task publishing takes a long time.

##### Key changed/added classes in this PR
 * `SeekableStreamSupervisor`
 * `SupervisorManager`
 * `SegmentTransactionalInsertAction`

This PR has:

- [X] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [ ] added documentation for new or modified features or behaviors.
- [ ] a release note entry in the PR description.
- [ ] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [X] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [ ] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [X] been tested in a test Druid cluster.

### CODE REVIEW

**Comment by kfaraz** [comment]
@georgew5656, this probably doesn't qualify as a bug per se.
It is a quality of life change but there is no incorrect behaviour here.

IIUC, this PR tries to:
- eagerly identify the conditions which result in the messages below.
- fail the commit in a retryable manner rather than failing the whole task.
> [Case A] The new start metadata state[%s] is ahead of the last committed" + " end state[%s]. Try resetting the supervisor
> [Case B] Inconsistency between stored metadata state[%s] and target state[%s]. Try resetting the supervisor.

If this is the case, we probably don't even need the new method `canPublishSegments`.
We just need to throw the right kind of exception for the above mentioned messages which can notify the task that it needs to retry. Or maybe the task action itself could retry. (something `segmentAllocate` action already does if the used segments seem to have changed).
However, if we do decide to have the `canPublishSegments` method, it should have the same logic that is used while validating the commit i.e. the same logic that ends up throwing the "inconsistent metadata" errors.


#### Alternatives

Perhaps, we could just lock/queue on the (supervisor + groupId) combo?
Publish should typically not take a long time, and failing every time some other task in the same group is already publishing seems like weird behaviour. The publish request coming from the second task could just wait while the first one finishes.

Let me know what you think.

**Comment by georgew5656** [comment]
> @georgew5656, this probably doesn't qualify as a bug per se. It is a quality of life change but there is no incorrect behaviour here.
> 
> IIUC, this PR tries to:
> 
> * eagerly identify the conditions which result in the messages below.
> * fail the commit in a retryable manner rather than failing the whole task.
> 
> > [Case A] The new start metadata state[%s] is ahead of the last committed" + " end state[%s]. Try resetting the supervisor
> > [Case B] Inconsistency between stored metadata state[%s] and target state[%s]. Try resetting the supervisor.
> 
> If this is the case, we probably don't even need the new method `canPublishSegments`. We just need to throw the right kind of exception for the above mentioned messages which can notify the task that it needs to retry. Or maybe the task action itself could retry. (something `segmentAllocate` action already does if the used segments seem to have changed). However, if we do decide to have the `canPublishSegments` method, it should have the same logic that is used while validating the commit i.e. the same logic that ends up throwing the "inconsistent metadata" errors.
> 
> #### Alternatives
> Perhaps, we could just lock/queue on the (supervisor + groupId) combo? Publish should typically not take a long time, and failing every time some other task in the same group is already publishing seems like weird behaviour. The publish request coming from the second task could just wait while the first one finishes.
> 
> Let me know what you think.

i thought about this approach but the concern is that it's possible for these error logs (inconsistency with metadata state) to be final error states, like if offsets are manually updated or something. we don't really want to retry in these scenarios

**Comment by kfaraz** [comment]
> i thought about this approach but the concern is that it's possible for these error logs (inconsistency with metadata state) to be final error states, like if offsets are manually updated or something. we don't really want to retry in these scenarios

Hmm, I don't think I have ever seen a case of people updating offsets manually. At best, someone might completely remove the entry for a certain datasource, in which case the flow would still work.

**Comment by georgew5656** [comment]
> > i thought about this approach but the concern is that it's possible for these error logs (inconsistency with metadata state) to be final error states, like if offsets are manually updated or something. we don't really want to retry in these scenarios
> 
> Hmm, I don't think I have ever seen a case of people updating offsets manually. At best, someone might completely remove the entry for a certain datasource, in which case the flow would still work.
> 
> And even for that, simply resetting the supervisor is advised (and preferred by users) rather than updating the DB directly.

i ran some more testing and i've discovered that the scenario described in this PR initially only triggers the "The new start metadata state[%s] is ahead of the last committed" message. (Later tasks can fail with the inconsistent metadata state message). Based on this I think its fine to just throw a retryable exception in the first case and let the mysql connector retry instead of passing it up to the task action level (it looks like it defaults to 10 retries).

I think this" new start metadata is ahead" state is relatively safe to retry as well, i believe the terminal failures states normally throw "inconsistent metadata" instead

**Comment by kfaraz** [comment]
The new approach is definitely much simpler.
@georgew5656 , IIRC, the 10 retries happen in an exponential backoff manner.
What is the total time that will be spent in 10 retries?
Will it be enough to allow the preceding publish to finish. (I think it should be).

File: server/src/main/java/org/apache/druid/metadata/IndexerSQLMetadataStorageCoordinator.java [hunk 1]
```diff
@@ -2661,7 +2661,7 @@ protected DataStoreMetadataUpdateResult updateDataSourceMetadataWithHandle(
 
     if (startMetadataGreaterThanExisting && !startMetadataMatchesExisting) {
       // Offsets stored in startMetadata is greater than the last commited metadata.
-      return DataStoreMetadataUpdateResult.failure(
+      return DataStoreMetadataUpdateResult.retryableFailure(
```
Comment (kfaraz) [inline server/src/main/java/org/apache/druid/metadata/IndexerSQLMetadataStorageCoordinator.java:2664]: Please add another line to the comment above clarifying why we should retry in this case.

**Comment by georgew5656** [comment]
> The new approach is definitely much simpler. @georgew5656 , IIRC, the 10 retries happen in an exponential backoff manner. What is the total time that will be spent in 10 retries? Will it be enough to allow the preceding publish to finish. (I think it should be).

its about a minute which i think should be enough in most cases. technically the previous approach also relied on client exponential backoff so we should be talking about a similar amount of time if the retry logic is the same

**Comment by kfaraz** [comment]
> its about a minute which i think should be enough in most cases

Looking at `RetryUtils`, I think the total time for 10 retries would be more than that, probably closer to 5 minutes or so.

> technically the previous approach also relied on client exponential backoff so we should be talking about a similar amount of time if the retry logic is the same

I think the retry logic is different since `ServiceClientImpl` uses the `StandardRetryPolicy` where the default exponential backoff leads to a total retry time of 1 hour approx.

That said, I prefer the retry at the metadata level since it is quicker and simpler.

**Review by kfaraz — APPROVED** [review]
The changes look good to me.

Could you please try to verify the time spent by 10 retries in the current patch so that we can be sure? We wouldn't want the Overlord to keep retrying the transaction for a long time.
I don't see much point in adding a UT/IT for this, verifying it in a local cluster should be sufficient.

**Pull request merged**